proposed/activity-improvements.md

```diff
@@ -0,0 +1,465 @@
+    /// Links can be used to represent batched operations where a Activity was initiated by multiple initiating Activities,
+    /// each representing a single incoming item being processed in the batch.
+    public readonly struct ActivityLink : IEquatable<ActivityLink>
+    {
+        public ActivityLink(ActivityContext context, IEnumerable<KeyValuePair<string, object>>? attributes = null);
+        public ActivityContext Context { get; }
+        public IEnumerable<KeyValuePair<string, object>>? Attributes { get; }
+        
+        public override bool Equals(object? obj);
+        public bool Equals(ActivityLink value);
+        public static bool operator ==(ActivityLink left, ActivityLink right);
+        public static bool operator !=(ActivityLink left, ActivityLink right);
+        public override int GetHashCode();
+    }
+}
+```
+
+### ActivityDataRequest
+
+```C#
+namespace System.Diagnostics
+{
+    /// Used by ActivityListener to indicate what amount of data should be collected for this Activity
+    /// Requesting more data causes greater performance overhead to collect it.
+    public enum ActivityDataRequest
+    {
+        /// The Activity object doesn't need to be created
+        None,
+
+        /// The Activity object needs to be created. It will have Name, Source, Id and Baggage.
+        /// Other properties are unnecessary and will be ignored by this listener.
+        PropagationData,
+
+        /// The activity object should be populated with all the propagation info and also all other
+        /// properties such as Links, Tags, and Events. Activity.IsAllDataRequested will return true.
+        AllData,
+
+        /// The activity object should be populated the same as the AllData case and additionally
+        /// Activity.IsRecorded is set true. For activities using W3C trace ids this sets a flag bit in the
+        /// ID that will be propagated downstream requesting that trace is recorded everywhere.
+        AllDataAndRecorded
+    }
+}
+```
+
+### ActivityCreationOptions
+
+```c#
+namespace System.Diagnostics
+{
+    public readonly struct ActivityCreationOptions<T>
+    {
+        public ActivitySource Source { get; }
+        public string Name { get; }
+        public ActivityKind Kind { get; }
+        public T Parent { get; }
+        public IEnumerable<KeyValuePair<string, string?>> Tags { get; }
+        public IEnumerable<ActivityLink> Links { get; }
+    }
+}
+```
+
+### ActivityListener
+
+```C#
+namespace System.Diagnostics
+{
+    public delegate ActivityDataRequest GetRequestedData<T>(ref ActivityCreationOptions<T> options);
+
+    public sealed class ActivityListener : IDisposable
+    {
+        public ActivityListener() { }
+        public Action<Activity> ActivityStopped { get; set; }
+        public Action<Activity> ActivityStarted { get; set; }
+        public Func<ActivitySource, bool> ShouldListenTo { get; set; }
```

Comment:
Does ActivityListener has the ability to enable/disable based on Activity name? Or is it only based on the ActivitySource name?

In the examples shared above, can ActivityListener say "I want to listen to Azure.Core.Http source, but only for activity named 'Request' "?

Reply:
The ActivityListener can listen to ActivitySource to get the Activity Start/Stop events. When any activity created by the ActivitySource get started or stopped, the listener will get called for this event and will have the Activity name passed to the listener callback. so, Listeners get registered by the ActivitySource and will get called later to filter according to the activity names.

Reply:
Got it thanks!